### PR TITLE
Fix MacOS retina support

### PR DESCRIPTION
**Pull request by khogeland**
Ensures the IOSurfaces are created with real pixel values.

See https://github.com/runelite/runelite/issues/17575

### CODE REVIEW

**Comment by Adam-** [comment]
> This breaks when the display resolution is changed with the display resolution is changed with the game open

Can you elaborate on what this means? Do you mean the resolution or the scaling factor?

**Comment by khogeland** [comment]
> > This breaks when the display resolution is changed with the display resolution is changed with the game open
> 
> Can you elaborate on what this means? Do you mean the resolution or the scaling factor?

Actually, scratch that, I can't repro anymore, it works fine when switching resolution.

**Comment by khogeland** [comment]
Ok, this _mostly_ works, but the screen will freak out when, very specifically: launching with a scale factor of 2, switching to a scale factor of 1, switching back to 2. The screen will flicker until the client is resized or the GPU plugin toggled.

Someone more familiar with this code should take a look, this was way more than I planned on doing. 😄

**Comment by khogeland** [comment]
Okay, I think this is done now. The flickering was because only one buffer was getting resized, and I'm now getting the current window's scale factor via AWT so it'll update when dragging between monitors etc.

**Comment by Adam-** [comment]
Have you looked at either https://developer.apple.com/documentation/quartzcore/calayer/1410801-rasterizationscale?language=objc or https://developer.apple.com/documentation/quartzcore/calayer/1410746-contentsscale?language=objc ? You might be able to avoid the upcall to get the scale if you can just read it from the calayer.

**Comment by khogeland** [comment]
Yeah, we have to update that field ourselves because the layer we create isn't automatically managed by a view.

File: rlawt.h [hunk 1]
```diff
@@ -62,6 +62,8 @@ typedef struct {
 
 	int offsetX;
 	int offsetY;
+
+	GLfloat backingScaleFactor;
```
Comment (Adam-) [inline rlawt.h]: this is `CGfloat`

File: rlawt_mac.m [hunk 2]
```diff
@@ -293,19 +296,30 @@ JNIEXPORT void JNICALL Java_net_runelite_rlawt_AWTContext_swapBuffers(JNIEnv *en
...
 
 	if (!ctx->buffer[ctx->back]
-		|| IOSurfaceGetWidth(ctx->buffer[ctx->back]) != ctx->layer.frame.size.width 
-		|| IOSurfaceGetHeight(ctx->buffer[ctx->back]) != ctx->layer.frame.size.height) {
-		if (!rlawtCreateIOSurface(env, ctx)) {
-			return;
-		}
+		|| IOSurfaceGetWidth(ctx->buffer[ctx->back]) != ctx->layer.frame.size.width
+		|| IOSurfaceGetHeight(ctx->buffer[ctx->back]) != ctx->layer.frame.size.height
+		|| ctx->backingScaleFactor != ctx->layer.superlayer.contentsScale) {
```
Comment (Adam-) [inline rlawt_mac.m]: Since the scales are tied to the buffers I would probably make a `CGfloat scale[2]` in `AWTContext` to avoid this double-swap

**Comment by Adam-** [comment]
This has been added in d5229d053af14578e5c9e6c42eab523ff22c1628. Note that your PR is recreating both IO surfaces each frame due to not scaling the dimensions used for the resize check.

**Pull request closed without merging**